File: src/pyodide/matches.py
"""Pulling the matches out of a score matrix, as Coda asks for it.

The sixth capability on the bridge, and the cheapest of the lot: the same two packages again,
and a matrix that has almost always just come off the NBLAST node — so the runtime is booted,
the wheel is installed, and this is a `runPython` of one file plus one call.

Three questions rather than three implementations of one, which is why the mode is a
parameter here and not three entry points. They differ in what comes back:

- **top** is rectangular — `n` matches per group, padded — so it crosses as two `(groups, n)`
  arrays with the shape beside them.
- **above** is ragged, so it crosses CSR-style: an `offsets` array plus one flat `indices`
  and one flat `values`.
- **count** is one number per group and crosses as one array.

**Why this is on the bridge at all** is worth stating, because at Coda's matrix sizes it does
not have to be. A 500 x 500 partial sort is microseconds of JavaScript, and fastcore's version
exists for the tens-of-gigabyte matrices a browser will never hold. What it buys instead is
*parity*: `percentage` means "within X% of this group's own best", not "the top X%", and
`skip_self` means the diagonal rather than a name comparison. Both are decisions somebody
would otherwise re-make slightly differently here, and a match table that disagrees with navis
by a rule nobody wrote down is worse than one that is a few milliseconds slower.

## Padding, and the two shapes it takes

`top_matches` fills a short row with `-1` in the indices and NaN in the values, exactly as
`nblast_knn` fills one with `-1` and `-inf`. It is left in place and dropped on the JavaScript
side for the same reason: dropping it here would lose the rectangular shape that makes the two
arrays readable, and carrying it through would put a match called -1 in front of somebody.
`matches_above` has no padding — a group with nothing above the cutoff is simply an empty
slice — which is the whole point of the CSR form.
"""

import numpy as np
import navis_fastcore as fc


def _cutoff(req):
    """The one of `threshold` / `percentage` this request names, as kwargs.

    `matches_above` and `count_matches` both want **exactly one** of the two and raise on
    neither or both, so the choice is made here rather than by passing both and hoping.
    """
    if str(req["cutoff"]) == "percentage":
        return {"percentage": float(req["percentage"])}
    return {"threshold": float(req["threshold"])}


def coda_matches_run(request, report=None):
    """Extract matches from a square or rectangular score matrix.

    The matrix arrives as a buffer and is read with `np.frombuffer`, which yields a
    **read-only** array. All three fastcore calls here take it without copying and none of
    them writes to it — checked rather than assumed, on `coda_linkage_run`'s finding about
    `linkage` versus `symmetrize`.
    """
    req = request.to_py()
    rows = int(req["rows"])
    cols = int(req["cols"])
    scores = np.frombuffer(req["scores"], dtype=np.float64).reshape(rows, cols)

    mode = str(req["mode"])
    axis = int(req["axis"])
    common = {
        "axis": axis,
        "distances": bool(req["distances"]),
        # `True` means the diagonal and needs a square matrix; the caller has already refused
        # the rectangular case, since "no self-match" on a query-versus-target comparison is a
        # question about names that a matrix cannot answer.
        "skip_self": bool(req["skipSelf"]),
    }
    groups = rows if axis == 0 else cols

    if report is not None:
        report(0.1, f"{rows} x {cols} matrix")

    if mode == "count":
        counts = np.asarray(fc.count_matches(scores, **_cutoff(req), **common))
        if report is not None:
            report(1.0, f"{groups} groups")
        return {
            "mode": "count",
            # int32 for the reason `coda_nblast_knn_run` casts its `idx`: fastcore counts in
            # int64 and an int64 numpy array crosses to JavaScript as a `BigInt64Array`, which
            # converts without complaint and then compares equal to nothing.
            "counts": np.ascontiguousarray(counts, dtype=np.int32),
            "groups": groups,
        }

    if mode == "top":
        idx, values = fc.top_matches(scores, int(req["n"]), **common)
        # Only `idx` is coerced here, and only because `idx.shape` is read below;
        # `ascontiguousarray` in the return dict accepts whatever `asarray` would have made of
        # the rest, so a second pass over them would allocate for nothing.
        idx = np.asarray(idx)
        if report is not None:
            report(1.0, f"{groups} x {idx.shape[1]} matches")
        return {
            "mode": "top",
            "idx": np.ascontiguousarray(idx, dtype=np.int32).ravel(),
            "values": np.ascontiguousarray(values, dtype=np.float64).ravel(),
            "groups": int(idx.shape[0]),
            "n": int(idx.shape[1]),
        }

    offsets, idx, values = fc.matches_above(
        scores,
        **_cutoff(req),
        max_matches=int(req["maxMatches"]),
        **common,
    )
    if report is not None:
        report(1.0, f"{len(idx)} matches")
    return {
        "mode": "above",
        "offsets": np.ascontiguousarray(offsets, dtype=np.int32),
        "idx": np.ascontiguousarray(idx, dtype=np.int32).ravel(),
        "values": np.ascontiguousarray(values, dtype=np.float64).ravel(),
        "groups": groups,
    }

File: src/pyodide/matches.ts
/**
 * Match extraction's typed wrapper over the Python bridge.
 *
 * The sixth capability, `nblast.ts`'s shape unchanged. The one thing worth knowing before
 * reading it is that the three modes return three different shapes, so the result is a
 * discriminated union rather than one type with everything optional — a `values` array that
 * is sometimes ranked-and-padded and sometimes ragged is two arrays wearing one name.
 */

import { callPython } from './engine'
import type { CallOptions } from './engine'
import { float64From, int32From, numberFrom } from './types'

/** Which question to ask of the matrix. */
export type MatchMode = 'top' | 'above' | 'count'

/** Which of the two cutoffs `above` and `count` are given. fastcore wants exactly one. */
export type MatchCutoff = 'threshold' | 'percentage'

export type MatchesRequest = {
  /**
   * Row-major `rows * cols`.
   *
   * **A copy, never the matrix off the wire** — `callPython` transfers every typed array in
   * a call's arguments, so passing an upstream value's own buffer detaches the scheduler's
   * cached result for the node above. `matchOps.ts` copies, exactly as `linkageOps.ts` does.
   */
  scores: Float64Array
  rows: number
  cols: number

  mode: MatchMode
  /** `0` groups by row (one set of matches per query); `1` groups by column. */
  axis: 0 | 1
  /** Lower is better. Derived from the matrix's own `measure` unless the node overrides it. */
  distances: boolean
  /** Exclude each group's diagonal cell. Needs a square matrix; the node refuses otherwise. */
  skipSelf: boolean

  /** Matches per group, for `top`. */
  n: number
  cutoff: MatchCutoff
  /** Absolute cutoff, for `above` and `count`. */
  threshold: number
  /** Band around each group's *own* best value, in `[0, 1]`. */
  percentage: number
  /** Refuse to allocate more than this many matches, for `above`. */
  maxMatches: number
}

/** `n` matches per group, rectangular and padded with `-1` / NaN. */
export interface TopMatchesResult {
  mode: 'top'
  /** Row-major `groups * n`, indices along the other axis, best first. */
  idx: Int32Array
  /** Aligned to `idx`; NaN where the index is `-1`. */
  values: Float64Array
  groups: number
  n: number
}

/** Everything clearing the cutoff, CSR-style: group `g` is `idx[offsets[g]..offsets[g+1]]`. */
export interface MatchesAboveResult {
  mode: 'above'
  offsets: Int32Array
  idx: Int32Array
  values: Float64Array
  groups: number
}

/** How many each group *would* yield, without materialising any of them. */
export interface MatchCountsResult {
  mode: 'count'
  counts: Int32Array
  groups: number
}

export type MatchesResult = TopMatchesResult | MatchesAboveResult | MatchCountsResult

/** Extract matches from a score matrix, in whichever of the three shapes was asked for. */
export async function runMatches(
  request: MatchesRequest,
  options: CallOptions = {},
): Promise<MatchesResult> {
  const result = await callPython(
    { module: 'matches', fn: 'coda_matches_run', args: [request] },
    options,
  )

  const groups = numberFrom(result, 'groups')

  if (request.mode === 'count') {
    const counts = int32From(result, 'counts')
    if (counts.length !== groups) {
      throw new Error(`Match counts returned ${counts.length} numbers for ${groups} groups`)
    }
    return { mode: 'count', counts, groups }
  }

  const idx = int32From(result, 'idx')
  const values = float64From(result, 'values')
  if (idx.length !== values.length) {
    throw new Error(`Matches returned ${idx.length} indices and ${values.length} scores`)
  }

  if (request.mode === 'top') {
    const n = numberFrom(result, 'n')
    if (idx.length !== groups * n) {
      throw new Error(`Top matches returned ${idx.length} entries for ${groups} x ${n}`)
    }
    return { mode: 'top', idx, values, groups, n }
  }

  const offsets = int32From(result, 'offsets')
  if (offsets.length !== groups + 1) {
    throw new Error(`Matches returned ${offsets.length} offsets for ${groups} groups`)
  }
  return { mode: 'above', offsets, idx, values, groups }
}
